[dyninst/compiler] Implement logical and physical encoding

piob-io opened the pull request:
### What does this PR do?

Implement compiler for translating ir.Program into eBPF binary.
Note that stack machine implementation needs adjustment to make everything work end-to-end,
next PR includes shim that makes necessary adjustment and tests that program can be successfully loaded (i.e. verified).

### Motivation
https://datadoghq.atlassian.net/browse/DEBUG-3887

Commit:
[dyninst/compiler/logical] Implement logical program encoding

Commit:
[dyninst/compiler/codegen] Implement physical program encoding

ajwerner reviewed (commented):
Here's my initial reaction:

It's not clear to me what the `compiler` package is going to do. Perhaps putting a doc comment on the package will help. Is it responsible only for vm encoding or will it also compile and output bpf object files and attachment metadata?

The `Encoder` feels a bit low-level to be exported. I'd have expected something that encapsulates the compiler's dependencies as the exported type that itself has some method like `Compile`. Like it'd be bogus to call `EncodeProgram` twice on the same encoder, but that's not captured in the API. Also, `EncodeLocationOp` feels like a weird thing to export. If this is all about generating a vm logical program I might call the object that does the generating a `Generator`,  have a constructor function for it that takes the dependencies, and then it'll have one method that takes an input program and outputs some representation of a logical program. I'd have that be in a separate package from the code that compile bpf.

I might organize this like `vm` for the stuff in `logical` and then a peer `vmgen` for generating the `vm` programs or perhaps more wordily, I could imagine:

```
vm/ -- raw interfaces and low-dependency definitions, could be all of logical
vm/vmgen -- the current "compiler"
```

piob-io commented:
Thank you, this is what I needed :)
The compiler package will grow bigger to encompass everything needed to get from IR to compiled BPF code + entrypoint list for loading. The encoder part is for "logical" encoding/generation (naming to be adjusted). I will add docs and I will be more careful about what's exported or not. Great point about panics, wasn't sure about dealing with them, recovering at public API boundary sounds best.

andreimatei commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -92,7 +94,129 @@ func (e *Encoder) EncodeType(t ir.Type) (logical.FunctionID, bool) {
+			// in different registers and/or stack (represented with multiple loclist pieces) are not padded.
+			// Consecutive pieces of data stored in the same loclist piece are padded (this only happens when
+			// the location is a stack, Go never packs multiple data pieces into same register).
+			layout := e.typeMemoryLayout(op.Variable.Type)
```
consider s/layout/layoutPieces

piob-io replied:
Done

andreimatei commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -92,7 +94,129 @@ func (e *Encoder) EncodeType(t ir.Type) (logical.FunctionID, bool) {
+			for _, locPiece := range loclist.Pieces {
+				paddedOffset := layout[layoutIdx].PaddedOffset
+				nextLayoutIdx := layoutIdx
+				for nextLayoutIdx < len(layout) && layout[nextLayoutIdx].PaddedOffset-paddedOffset < uint32(locPiece.Size) {
```
Comment something like
```
// Advance nextLayoutIdx to the beginning of the next layout piece.
// Layout pieces in [layoutIdx, nextLayoutIdx) range correspond to current locPiece.
```

andreimatei commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -92,7 +94,129 @@ func (e *Encoder) EncodeType(t ir.Type) (logical.FunctionID, bool) {
+			// the location is a stack, Go never packs multiple data pieces into same register).
+			layout := e.typeMemoryLayout(op.Variable.Type)
+			layoutIdx := 0
+			outputOffset := op.Offset
```
Please comment what `outputOffset` tracks.

piob-io replied:
As above

andreimatei commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -69,29 +84,162 @@ func (e *Encoder) EncodeExpression(subprogram *ir.Subprogram, injectionPC uint64
-func (e *Encoder) EncodeType(t ir.Type) (logical.FunctionID, bool) {
-	panic("not implemented")
+func (e *Encoder) encodeRemainingTypes() {
+	for i := 0; i < len(e.typeQueue); i++ {
```
nit: consider writing this more in a `while` style, to suggest that elements are appended to `typeQueue` while iterating (I think the loop is correct the way you wrote it, but I think replacing it with a `range` would not be valid).
So
```
for len(queue) > 0 {
   encodeType()
   queue = queue[1:]
}
```

piob-io replied:
Done

andreimatei commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -17,20 +17,35 @@ import (
 	typeResolver TypeResolver
 
-	Functions map[logical.FunctionID][]logical.Op
+	// Queue of interesting types that need a `ProcessType` function.
```
maybe add more words explaining when types are discovered and when they're consumed from this queue.
Better yet, consider whether this deserves to be a file on the `Encoder`, as opposed to passing it around as arguments to the few functions that need to append to the queue -- then we wouldn't need to talk about comments on it :P

piob-io replied:
I changed Encoder to be internal only type that tracks encoding. I think it's simpler than passing around callbacks, but I could do that as well. As for the comment, I do prefer the code to explain itself, is it enough now that the encoder is private and all uses of this are internal?

andreimatei replied:
I wasn't suggesting any callbacks; I was suggesting that this queue could be local to `EncodeProgram` and passed as an argument to `encodeType()`. But I no longer think that's necessarily better now that `EncodeProgram` is not a method on `encoder`.

andreimatei commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -69,29 +84,162 @@ func (e *Encoder) EncodeExpression(subprogram *ir.Subprogram, injectionPC uint64
 
-func (e *Encoder) EncodeType(t ir.Type) (logical.FunctionID, bool) {
-	panic("not implemented")
+func (e *Encoder) encodeRemainingTypes() {
```
consider inlining this function in the (single) caller. Otherwise I'd probably nitpick about explaining the word "remaining".

piob-io replied:
Done

andreimatei commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -69,29 +84,162 @@ func (e *Encoder) EncodeExpression(subprogram *ir.Subprogram, injectionPC uint64
+}
+
+// Generate `ProcessType` function, if needed for a given type.
+func (e *Encoder) encodeType(t ir.Type) (logical.FunctionID, bool) {
```
Please comment the return values. Can you explain when the function is not needed?

piob-io replied:
Done. I added an example. Although this is to the implementation to decide, and caller responsibility is that if it loaded data of a given type, and there if a function for that type, it should call it.

andreimatei commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -69,29 +84,162 @@ func (e *Encoder) EncodeExpression(subprogram *ir.Subprogram, injectionPC uint64
+		needed = true
+		offsetShift = uint32(t.GetByteSize())
+		ops = []logical.Op{
+			logical.ProcessArrayPrep{},
```
Do we really need `ProcessArrayPrep/Repeat` and `ProcessSlicePrep/Repeat` (and maybe others?) instead of a more generic looping mechanism?

piob-io replied:
We don't need it. But I do think this is better than what we had before:
https://github.com/DataExMachina-dev/side-eye/blob/f0854ec10f9db7772b648db4eff97fe91df651cf/crates/probe-vm-gen/src/generate/enqueue.rs#L163-L182

The reasoning is, this functions all manipulate state of the stack machine. And when something goes wrong when executing, it's easier to debug when you are closer to that state. When we have a lot of instructions, there are two complex levels, understanding what all of the instructions are supposed to do between each other (emergent property of having stack machine code), and what each instruction is actually doing (direct property of the instruction implementation).

If we had lots of tooling invested into analyzing stack machine code, evaluators, debuggers, etc, then I would go with lower level instructions. But given that we are aiming at light-weight, narrow-purpose stack machine, I prefer to go with higher level instructions.

andreimatei replied:
sg

andreimatei commented on pkg/dyninst/compiler/logical/functions.go:
```diff
@@ -44,6 +44,7 @@ type ProcessExpression struct {
 }
 
 // Process user data of a specific type, chasing pointers, resolving interfaces, etc (after the data was already read into ringbuf).
+// Output offset must be set to the beginning of the data. The function implementation must advance the offset to the end of the data.
```
I think it'd help the casual reader if you'd say "The generated function expects the output offset to be set...", to make it clear that we're talking at a meta level.

andreimatei reviewed (commented):
mostly asking for comments :)

piob-io commented on pkg/dyninst/compiler/logical/deps.go:
```diff
@@ -0,0 +1,12 @@
+import "github.com/DataDog/datadog-agent/pkg/dyninst/ir"
+
+type TypeResolver interface {
+	Resolve(id ir.TypeID) ir.Type
```
Done

piob-io commented on pkg/dyninst/compiler/logical/functions.go:
```diff
@@ -0,0 +1,66 @@
+	TypeID ir.TypeID
+}
+
+func (ChasePointers) PrettyString(resolver TypeResolver) string {
```
Done

piob-io commented on pkg/dyninst/compiler/logical/ops.go:
```diff
@@ -0,0 +1,135 @@
+// program counter back to itself, effectively looping. They can also use data stack
+// to control this behavior, for example to track loop counter.
+
+type CallOp struct {
```
Let's chat more offline

piob-io commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -7,20 +7,92 @@
+	}
+}
+
+func (e *Encoder) EncodeEvent(subprogram *ir.Subprogram, injectionPC uint64, rootType *ir.EventRootType) {
```
Maybe? Although I don't follow indication that "event" is the output... Why? Event is something that happens, a trigger, that we have handling for here...

piob-io replied:
Oh, yeah "Add" is better, and so AddEventHandler sg. Also addExpressionHandler and addTypeHandler. Commented on the goals of the generated functions.

piob-io commented on pkg/dyninst/compiler/logical/functions.go:
```diff
@@ -31,36 +31,36 @@ type ChasePointers struct {
-func (e ProcessExpression) PrettyString(resolver TypeResolver) string {
-	return fmt.Sprintf("ProcessEvent[%s@%x.expr[%d]]", resolver.Resolve(e.EventRootTypeID).GetName(), e.InjectionPC, e.ExprIdx)
+func (e ProcessExpression) PrettyString() string {
+	return fmt.Sprintf("ProcessEvent[%s@%x.expr[%d]]", e.EventRootType.GetName(), e.InjectionPC, e.ExprIdx)
```
Done

piob-io commented on pkg/dyninst/compiler/logical/functions.go:
```diff
@@ -31,36 +31,36 @@ type ChasePointers struct {
-	ExprIdx         int
-	InjectionPC     uint64
+	EventRootType *ir.EventRootType
+	ExprIdx       int
```
We need the number, for example to know which presence bit in the presence bitset to set. Commented.

piob-io commented on pkg/dyninst/compiler/logical/ops.go:
```diff
@@ -64,6 +66,8 @@ type ExprPrepareOp struct {
 type ExprSaveOp struct {
 	baseOp
+	EventRootType *ir.EventRootType
+	ExprIdx       int
```
Commented

piob-io commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -7,20 +7,92 @@
+	e.AddFunction(id, ops)
+}
+
+func (e *Encoder) EncodeExpression(subprogram *ir.Subprogram, injectionPC uint64, rootType *ir.EventRootType, exprIdx int) logical.FunctionID {
```
Commented on the params. Removed subprogram, it's not needed here (expressions directly point to relevant variables).

piob-io commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -92,7 +94,129 @@ func (e *Encoder) EncodeType(t ir.Type) (logical.FunctionID, bool) {
+			}
+		}
+	}
+	// Variable is not available. Expression ops should "return early" on error.
```
There was a missing return in the loop, we only come here if we didn't find the relevant pc range. Rephrased to indicate that just returning is ok.

piob-io commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -7,20 +7,92 @@
+	panic("not implemented")
+}
+
+// Appends encoded instructions to `ops`.
```
Done, ptal

piob-io commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -92,7 +94,129 @@ func (e *Encoder) EncodeType(t ir.Type) (logical.FunctionID, bool) {
+}
+
+// TODO: move padded offset calculation for location pieces into IR generation.
+// Breaks down a type memory into (data size, padding size) pairs. Padding size may be zero, and consecutive
```
Will do (not done yet)

piob-io commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -92,7 +94,129 @@ func (e *Encoder) EncodeType(t ir.Type) (logical.FunctionID, bool) {
+func (e *Encoder) typeMemoryLayout(t ir.Type) []memoryLayoutPiece {
+	pieces := make([]memoryLayoutPiece, 0)
+	var collectPieces func(t ir.Type, offset uint32)
+	collectPieces = func(t ir.Type, offset uint32) {
```
Fair, I will apply this when resolving TODO about moving this code into IR (this code will work slightly differently there, and I think will also be a bit simpler).

piob-io commented on pkg/dyninst/compiler/encoder.go:
```diff
@@ -92,7 +94,129 @@ func (e *Encoder) EncodeType(t ir.Type) (logical.FunctionID, bool) {
+func (e *Encoder) typeMemoryLayout(t ir.Type) []memoryLayoutPiece {
+	pieces := make([]memoryLayoutPiece, 0)
+	var collectPieces func(t ir.Type, offset uint32)
+	collectPieces = func(t ir.Type, offset uint32) {
```
As above, will improve when moving.

andreimatei commented on pkg/dyninst/compiler/logical/encode.go:
```diff
@@ -0,0 +1,387 @@
+// `ops` is used as an output buffer for the encoded instructions.
+func (e *encoder) EncodeLocationOp(pc uint64, op *ir.LocationOp, ops []Op) []Op {
+	for _, loclist := range op.Variable.Locations {
+		if loclist.Range.Start <= pc && pc < loclist.Range.End {
```
nit: I think this would read a lot better if we unindented the happy case.
```
for _, loclist := range op.Variable.Locations {
  if pc < loclist.Range.Start || pc > loclist.Range.End {
    continue
  }
  ...
  return ops
}
// We didn't find...
return append(ops, ReturnOp{})
```

piob-io replied:
Done

andreimatei commented on pkg/dyninst/compiler/logical/encode.go:
```diff
@@ -0,0 +1,387 @@
+			return ops
+		}
+	}
+	// Variable is not available, just return. Expression ops are allowed to "return early" on error.
```
I still don't quite understand this comment. Some ops are allowed to return early on error, but what if there's no error -- who returns then?

piob-io replied:
Also in this specific case we would produce just the return op and stop, without generating any following, unreachable instructions.

piob-io commented on pkg/dyninst/compiler/logical/encode.go:
```diff
@@ -0,0 +1,387 @@
+	e.addFunction(id, ops)
+}
+
+// InjectionPC is the PC of the user program at which the expressions are evaluated.
```
Done

ajwerner commented on pkg/dyninst/compiler/logical/encode.go:
```diff
@@ -0,0 +1,406 @@
+	Types     []ir.Type
+}
+
+type encoder struct {
```
nit: again I don't know about the `encode` terminology.

piob-io replied:
Changed

ajwerner reviewed (commented):
High level seems fine. Lots of little comments but nothing major. One thing I wonder about is whether readability and clarity around what sort of C code is being generated would be improved through the use of the https://pkg.go.dev/text/template package. I think there's a refactoring that uses that thing and lays out much more clearly what the generated C will look like.

Commit:
[dyninst/compiler] Reorganize logical/physical packages into sm and codegen.

Sm deals with both logical and physical representation of the stack machine program.
Codegen deals with C code generation.

piob-io commented on pkg/dyninst/compiler/logical/encode.go:
```diff
@@ -0,0 +1,406 @@
+}
+
+// EncodeProgram encodes the IR program into a logical eBPF stack machine program.
+func EncodeProgram(program ir.Program) Program {
```
Yup, went with sm package and GenerateProgram producing sm.Program.

piob-io commented on pkg/dyninst/compiler/codegen/code_test.go:
```diff
@@ -0,0 +1,64 @@
+			},
+		},
+	}
+	GenerateCode(logical.Program{Functions: functions}, os.Stdout)
```
Let me to do that in a separate PR, I do plan to have some golden-style test that shows generated code.

piob-io commented on pkg/dyninst/compiler/logical/functions.go:
```diff
@@ -0,0 +1,71 @@
+	"github.com/DataDog/datadog-agent/pkg/dyninst/ir"
+)
+
+type FunctionID interface {
```
Ha, gtk, done

piob-io commented on pkg/dyninst/compiler/codegen/code.go:
```diff
@@ -0,0 +1,136 @@
+// Each code fragment must be able to declare apriori how many bytes it will
+// generate.
+type encodable interface {
+	codeByteLen() uint32
```
Renamed, added comments and restructured into sm package and codegen, first dealing with abstract stack machine, second dealing with C code. Thanks for the conversation.

piob-io commented on pkg/dyninst/compiler/logical/ops.go:
```diff
@@ -0,0 +1,149 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package logical
```
Done

piob-io commented on pkg/dyninst/compiler/logical/functions.go:
```diff
@@ -0,0 +1,71 @@
+	baseFunctionID
+}
+
+func (ChasePointers) PrettyString() string {
```
Wasn't sure what was the convention, wanted to imply it's human readable. Changed.

piob-io commented on pkg/dyninst/compiler/logical/encode.go:
```diff
@@ -0,0 +1,402 @@
+// may be zero, and consecutive data pieces with zero padding may not be
+// coalesced. Only supports data that may be stored in registers and on stack.
+func (e *encoder) typeMemoryLayout(t ir.Type) []memoryLayoutPiece {
+	pieces := make([]memoryLayoutPiece, 0)
```
I see, gtk, changed.

piob-io commented on pkg/dyninst/compiler/codegen/code.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+// GenerateCode generates:
+//   - stack machine code
```
Done, it's also clearer with new package structure.

piob-io commented on pkg/dyninst/compiler/codegen/code.go:
```diff
@@ -0,0 +1,136 @@
+type tracker struct {
+	// PC of the first instruction of the function, used for call ops.
+	functionLoc map[logical.FunctionID]uint32
+}
```
Encode does not and should not mutate the tracker. I hope it's clearer with reorganized code and comments. So I prefer to keep it passed by value. And I want it to be a struct, even though it's a single type, because it lets the code be a bit more self-descriptive. And for performance reasons, even slightly bigger structures are cheaper to copy compare to memory costs of pointer hopping.

Commit:
[dyninst/compiler] Error instead of panic on non statically checked invariants

piob-io commented on pkg/dyninst/compiler/logical/encode.go:
```diff
@@ -0,0 +1,402 @@
+		case *ir.LocationOp:
+			ops = e.EncodeLocationOp(injectionPC, op, ops)
+		default:
+			panic(fmt.Sprintf("unexpected ir.Operation: %#v", op))
```
Thanks for offline discussion. I changed most of panics into errors, even for internal invariants that aren't statically checked. The only remaining panics are for type switches that are supposed to be exhaustive, with panic for default case.

ajwerner reviewed (commented):
I'm fine with this. I find this lazy conversion from sm.Op to `instruction`/`codeFragment` while doing serialization to be a bit too cute for my taste, but certainly not so much so that I'm going to push back on it.

I suppose that in theory this streaming approach lets you avoid materializing all the instructions at the same time in a non-serialized form, and additionally you could avoid allocations on a per-instruction basis, but that latter optimization is not the case in this implementation.

piob-io commented on pkg/dyninst/compiler/codegen/code.go:
```diff
@@ -0,0 +1,66 @@
+)
+
+// CCodeSerializer serializes the stack machine code into C code.
+type CCodeSerializer struct {
```
Done

piob-io commented:
> I'm fine with this. I find this lazy conversion from sm.Op to `instruction`/`codeFragment` while doing serialization to be a bit too cute for my taste, but certainly not so much so that I'm going to push back on it.
> 
> I suppose that in theory this streaming approach lets you avoid materializing all the instructions at the same time in a non-serialized form, and additionally you could avoid allocations on a per-instruction basis, but that latter optimization is not the case in this implementation.

I prefer lazy encoding, because then I don't need to rewrite call ops with the correct call pc. This way the logic to end up with correct call op bytes is closer to call op encoding implementation.

ajwerner reviewed (approved):
LGTM

piob-io commented:
/merge
